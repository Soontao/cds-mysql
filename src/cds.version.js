
const VERSION = "~7.8.0";
const MTXS_VERSION = "~1.17.0";

module.exports = { VERSION, MTXS_VERSION };
