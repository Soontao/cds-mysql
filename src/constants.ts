export const ANNOTATION_CDS_TYPEORM_CONFIG = "@cds.typeorm.config";

export const MYSQL_DATE_TIME_FORMAT = "yyyy-MM-dd HH:mm:ss.u";
export const MYSQL_DATE_TIME_FORMAT_WO_FRACTIONS = "yyyy-MM-dd HH:mm:ss";
export const MYSQL_CHARSET = "utf8mb4";
export const MYSQL_COLLATE = "utf8mb4_unicode_ci";
export const TENANT_DEFAULT = "default";

/**
 * the interval between idle connection check
 */
export const CONNECTION_IDLE_CHECK_INTERVAL = 5 * 1000;

/**
 * the default maximum connections number for each tenant
 */
export const DEFAULT_TENANT_CONNECTION_POOL_SIZE = 20;

/**
 * connection acquire timeout
 */
export const DEFAULT_CONNECTION_ACQUIRE_TIMEOUT = 10 * 1000;

/**
 * max queue items number to avoid over consuming for memory
 */
export const MAX_QUEUE_SIZE = 1000;

/**
 * the default connection idle timeout, after idle, it will be disconnected after next check
 */
export const DEFAULT_CONNECTION_IDLE_TIMEOUT = 120 * 1000;

/**
 * default value to check csv record is existed or not
 */
export const DEFAULT_CSV_IDENTITY_CONCURRENCY = 10;

/**
 * default `max_allowed_packet` value for `cds-mysql`
 * 
 * @see https://dev.mysql.com/doc/refman/8.0/en/packet-too-large.html
 */
export const DEFAULT_MAX_ALLOWED_PACKED_MB = 512;

export const MIGRATION_VERSION_PREFIX = "-- version: ";