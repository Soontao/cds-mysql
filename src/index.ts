import { MySQLDatabaseService } from "./service";

export = MySQLDatabaseService