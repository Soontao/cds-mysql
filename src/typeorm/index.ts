import { csnToEntity } from "./entity";
import { migrate, migrateData } from "./migrate";

export { entitySchemaToTable } from "./database";
export { csnToEntity, migrate, migrateData, };
