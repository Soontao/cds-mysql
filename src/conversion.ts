// conversion for OUTPUT (DB -> JSON)
import { cwdRequire, cwdRequireCDS } from "cds-internal-tool";
import { DateTime, FixedOffsetZone, ToISOTimeOptions } from "luxon";
import { MYSQL_DATE_TIME_FORMAT, MYSQL_DATE_TIME_FORMAT_WO_FRACTIONS } from "./constants";


/**
 * parse mysql date string
 * 
 * @param value valid mysql date string, with fractions or not
 * 
 * @returns 
 */
export function parseMysqlDate(value: string): DateTime;
export function parseMysqlDate(value: any): null;
export function parseMysqlDate(value: string) {
  if (typeof value !== "string") {
    return null;
  }

  switch (value.length) {
    case MYSQL_DATE_TIME_FORMAT_WO_FRACTIONS.length:
      return DateTime.fromFormat(
        value,
        MYSQL_DATE_TIME_FORMAT_WO_FRACTIONS,
        { zone: FixedOffsetZone.utcInstance }
      );
    default:
      return DateTime.fromFormat(
        value,
        MYSQL_DATE_TIME_FORMAT,
        { zone: FixedOffsetZone.utcInstance }
      );
  }
}


const convertToBoolean = (boolean: any) => {
  if (boolean === null) {
    return null;
  }

  return Boolean(boolean);
};

const convertToDateString = value => {
  if (!value) {
    return null;
  }

  return value;
};

const convertInt64ToString = int64 => {
  if (int64 === null) {
    return null;
  }

  return String(int64);
};

const createConvertToISOTime = (options?: ToISOTimeOptions) => (value: string) => {
  if (value === null || value === undefined) {
    return null;
  }
  const dateTime = parseMysqlDate(value);
  if (dateTime.isValid) {
    return dateTime.toISO(options);
  }
  return null;
};

const convertToISOTime = createConvertToISOTime();

const convertToISONoMilliseconds = createConvertToISOTime({ suppressMilliseconds: true });

const TYPE_POST_CONVERSION_MAP = new Map([
  ["cds.Boolean", convertToBoolean],
  ["cds.Date", convertToDateString],
  ["cds.Integer64", convertInt64ToString],
  ["cds.Int64", convertInt64ToString],
  ["cds.DateTime", convertToISONoMilliseconds],
  ["cds.Timestamp", convertToISOTime],
]);

if (cwdRequireCDS().env.features["bigjs"] === true) {
  const Big = cwdRequire("big.js");
  const convertToBig = (value: any) => new Big(value);

  TYPE_POST_CONVERSION_MAP.set("cds.Integer64", convertToBig);
  TYPE_POST_CONVERSION_MAP.set("cds.Int64", convertToBig);
  TYPE_POST_CONVERSION_MAP.set("cds.Decimal", convertToBig);
}

export { TYPE_POST_CONVERSION_MAP };

