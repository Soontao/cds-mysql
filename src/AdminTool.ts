/* eslint-disable max-len */
import { CSN, cwdRequireCDS, LinkedModel, Logger, memorized } from "cds-internal-tool";
import { DataSource, DataSourceOptions } from "typeorm";
import { TENANT_DEFAULT } from "./constants";
import { MysqlDatabaseOptions } from "./types";
import { formatTenantDatabaseName } from "./tenant";
import { migrate, migrateData } from "./typeorm";
import { csnToEntity } from "./typeorm/entity";
import { TypeORMLogger } from "./typeorm/logger";
import { CDSMySQLDataSource } from "./typeorm/mysql";

/**
 * get raw CSN from linked model
 */
export const _rawCSN = memorized(async (m: LinkedModel) => {
  return await cwdRequireCDS().load(m["$sources"]);
});

/**
 * Admin Tool for Database
 */
export class AdminTool {

  private _logger: Logger;

  private _options: MysqlDatabaseOptions;

  constructor() {
    this._logger = cwdRequireCDS().log("db|mysql");
    this._options = cwdRequireCDS().requires.db;
  }

  /**
   * get the database name of tenant
   * @param tenant 
   * @returns 
   */
  public getTenantDatabaseName(tenant: string = TENANT_DEFAULT) {
    // TODO: mysql have max length restriction for database
    return formatTenantDatabaseName(
      this._options?.credentials,
      this._options?.tenant?.prefix,
      tenant,
    );
  }

  /**
   * get mysql connection credential
   * 
   * @param tenant 
   * @returns 
   */
  public getMySQLCredential(tenant: string): import("mysql2").ConnectionOptions {
    return {
      ...this._options.credentials,
      dateStrings: true,
      database: this.getTenantDatabaseName(tenant),
    } as any;
  }

  /**
   * create database for tenant
   * 
   * @param tenant tenant id
   * @returns 
   */
  public async createDatabase(tenant: string): Promise<void> {
    if (tenant === undefined || tenant === TENANT_DEFAULT) {
      this._logger.debug("default tenant, skip creation database");
      return;
    }

    if (await this.hasTenantDatabase(tenant)) {
      this._logger.info(
        "try to create database for tenant",
        tenant.green,
        "but its already existed"
      );
      return;
    }

    await this.runWithAdminConnection(async ds => {
      const databaseName = this.getTenantDatabaseName(tenant);
      this._logger.info("creating database", databaseName);
      // mysql 5.6 not support 'if not exists'
      await ds.query(`CREATE DATABASE ${databaseName}`);
      this._logger.info("database", databaseName, "created");
    });

  }

  /**
   * drop database for tenant
   * 
   * @param tenant 
   */
  public async dropDatabase(tenant: string) {
    this._logger.info("drop database for tenant", tenant);

    if (tenant === undefined) {
      throw new Error("tenant id must be provided for database drop");
    }

    await this.runWithAdminConnection(async ds => {
      const databaseName = this.getTenantDatabaseName(tenant);
      this._logger.info("drop database", databaseName);
      await ds.query(`DROP DATABASE ${databaseName}`);
      this._logger.info("drop database", databaseName, "successful");
    });
  }

  /**
   * get csn for tenant
   * 
   * @param tenant 
   * @returns 
   */
  public async csn4(tenant?: string) {
    const { "cds.xt.ModelProviderService": mp } = cwdRequireCDS().services;
    return (mp as any).getCsn({ tenant, toggles: ["*"], activated: true });
  }

  /**
   * get typeorm ds option for migration or other usage
   * 
   * @param tenant 
   * @returns 
   */
  public async getDataSourceOption(tenant: string = TENANT_DEFAULT): Promise<DataSourceOptions> {
    const credentials = await this.getMySQLCredential(tenant);
    return Object.assign(
      {},
      {
        name: `cds-deploy-connection-${tenant ?? "main"}`,
        type: "mysql",
        entities: []
      },
      credentials,
      {
        // typeorm need the 'username' field as username
        username: credentials.user
      }
    ) as any;
  }

  /**
   * run admin operations
   * 
   * NOTICE, there is no pool for this queries
   * 
   * CAREFULLY use this please
   * 
   * @param runner 
   * @returns the returned value of runner 
   */
  public async runWithAdminConnection<T = any>(runner: (ds: DataSource) => Promise<T>): Promise<T> {
    const credential = await this.getDataSourceOption();
    const ds = new CDSMySQLDataSource({
      ...credential,
      name: `admin-conn-${cwdRequireCDS().utils.uuid()}`,
      entities: [],
      type: "mysql",
      logger: TypeORMLogger,
      synchronize: false, // no sync
    } as any);

    try {
      await ds.initialize();
      return await runner(ds);
    }
    catch (err) {
      this._logger.error("run with admin connection failed", err);
      throw err;
    }
    finally {
      if (ds && ds.isInitialized) {
        await ds.destroy();
      }
    }
  }

  /**
   * check tenant database is existed or not
   * 
   * @param tenant 
   * @returns 
   */
  public async hasTenantDatabase(tenant?: string) {
    // TODO: maybe cache for seconds
    return this.runWithAdminConnection(async ds => {
      const tenantDatabaseName = this.getTenantDatabaseName(tenant);
      const [{ COUNT }] = await ds.query(
        "SELECT COUNT(*) AS COUNT FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = ?",
        [tenantDatabaseName]
      );
      return COUNT > 0;
    });

  }

  /**
   * sync tenant data model
   * 
   * @param tenant 
   * @param csn CSN, optional 
   * @returns 
   */
  public async syncTenant(tenant: string, csn?: CSN) {
    if (csn !== undefined) {
      await this.deploy(csn, tenant);
      return;
    }

    // if has tenant database & mtxs is enabled
    if (await this.hasTenantDatabase(tenant) && ("cds.xt.ModelProviderService" in cwdRequireCDS().services)) {
      await this.deploy(await this.csn4(tenant), tenant);
      return;
    }

    await this.deploy(await _rawCSN(cwdRequireCDS().db.model), tenant);
  }

  /**
   * deploy CSV for tenant
   * 
   * @param tenant 
   */
  public async deployCSV(tenant?: string, csvList?: Array<string>) {
    // REVISIT: global model not suitable for extensibility
    await migrateData(this.getMySQLCredential(tenant), cwdRequireCDS().model, csvList);
  }

  /**
   * deploy (migrate) schema to (tenant) database
   * 
   * @param model plain CSN object
   * @param tenant tenant id
   * @returns 
   */
  public async deploy(model: CSN, tenant: string = TENANT_DEFAULT) {
    try {
      const cds = cwdRequireCDS();
      this._logger.info("migrating schema for tenant", tenant.green);
      if (tenant !== TENANT_DEFAULT) { await this.createDatabase(tenant); }
      const entities = csnToEntity(model);
      const migrateOptions = await this.getDataSourceOption(tenant);
      await migrate({ ...migrateOptions, entities });
      if (cds.env?.get?.("requires.db.csv.migrate") !== false) {
        await this.deployCSV(tenant);
      }
      else {
        this._logger.info(
          "csv migration is disabled for tenant",
          tenant.green
        );
      }
      this._logger.info("migrate", "successful".green, "for tenant", tenant.green);
      return true;
    } catch (error) {
      this._logger.info("migrate", "failed".red, "for tenant", tenant.red, error);
      throw error;
    }
  }

  /**
   * get tenant 0 (admin tenant) name
   *
   * @returns
   */
  getAdminTenantName() {
    return process.env.CDS_REQUIRES_MULTITENANCY_T0 ?? "t0";
  }

  /**
   * get tables of tenant
   * 
   * @param tenant 
   * @returns 
   */
  async getTables(tenant?: string): Promise<Array<string>> {
    return this.runWithAdminConnection(async ds => {
      const tables = await ds.query(
        `SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE' AND TABLE_SCHEMA = ?`,
        [this.getTenantDatabaseName(tenant)],
      );
      return tables.map(({ TABLE_NAME }) => String(TABLE_NAME).toLowerCase());
    });
  }

  /**
   * deploy admin tenant if required
   * 
   * @returns 
   */
  async deployT0() {
    const t0 = this.getAdminTenantName();
    // TODO: additional CSN configuration
    const csn = await cwdRequireCDS().load(`${__dirname}/../mtxs/t0.cds`);
    this._logger.info("deploy admin tenant", t0.green);
    await this.deploy(csn, t0);
  }

}