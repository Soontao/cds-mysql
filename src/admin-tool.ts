/* eslint-disable max-len */
import { CSN, cwdRequireCDS, LinkedModel } from "cds-internal-tool";
import path from "path";
import { DataSource, DataSourceOptions } from "typeorm";
import { TENANT_DEFAULT } from "./constants";
import { formatTenantDatabaseName } from "./tenant";
import { migrate, migrateData } from "./typeorm";
import { csnToEntity } from "./typeorm/entity";
import { TypeORMLogger } from "./typeorm/logger";
import { CDSMySQLDataSource } from "./typeorm/mysql";
import { MysqlDatabaseOptions } from "./types";
import fs from "fs/promises";
import { migration_tool } from "./utils";

const cds = cwdRequireCDS();
const logger = cds.log("admin-tool|db|mysql");

function _options() {
  return cds.env.requires.db as MysqlDatabaseOptions;
}

export function getTenantDatabaseName(tenant: string = TENANT_DEFAULT) {
  const options = _options();
  const tenantDatabaseName = formatTenantDatabaseName(
    options?.credentials,
    options?.tenant?.prefix,
    tenant,
  );

  if (tenantDatabaseName.length > 64) {
    logger.warn("database name", tenantDatabaseName, "which for tenant", tenant, "is too long");
    throw cds.error("TENANT_DATABASE_NAME_TOO_LONG");
  }

  return tenantDatabaseName;
}

export function getMySQLCredential(tenant: string): import("mysql2").ConnectionOptions {
  return {
    ..._options().credentials,
    dateStrings: true,
    database: getTenantDatabaseName(tenant),
  } as any;
}

export async function getDataSourceOption(
  tenant: string = TENANT_DEFAULT
): Promise<DataSourceOptions> {
  const credentials = getMySQLCredential(tenant);
  return Object.assign(
    {},
    {
      name: `cds-deploy-connection-${tenant ?? "main"}`,
      type: "mysql",
      entities: []
    },
    credentials,
    {
      // typeorm need the 'username' field as username
      username: credentials.user
    }
  ) as any;
}

export async function csn4(tenant?: string) {
  const { "cds.xt.ModelProviderService": mp } = cds.services;
  return mp.getCsn({ tenant, toggles: ["*"], activated: true });
}

export async function runWithAdminConnection<T = any>(runner: (ds: DataSource) => Promise<T>): Promise<T> {
  const credential = await getDataSourceOption();
  const ds = await CDSMySQLDataSource.createDataSource({
    ...credential,
    name: `admin-conn-${cds.utils.uuid()}`,
    entities: [],
    type: "mysql",
    logger: TypeORMLogger,
    synchronize: false, // no sync
  } as any);

  try {
    return await runner(ds);
  }
  catch (err) {
    logger.error("run with admin connection failed", err);
    throw err;
  }
  finally {
    if (ds && ds.isInitialized) {
      await ds.destroy();
    }
  }
}


/**
 * get raw CSN from linked model
 * 
 * cannot be cached will cause stack overflow
 */
export const _rawCSN = async (m: LinkedModel) => {
  return cds.load(m["$sources"]);
};

/**
 * check tenant database is existed or not
 * 
 * @param tenant 
 * @returns 
 */
export async function hasTenantDatabase(tenant?: string) {
  // TODO: maybe cache for seconds
  return runWithAdminConnection(async ds => {
    const tenantDatabaseName = getTenantDatabaseName(tenant);
    const [{ COUNT }] = await ds.query(
      "SELECT COUNT(*) AS COUNT FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = ?",
      [tenantDatabaseName]
    );
    return COUNT > 0;
  });

}

export async function createDatabase(tenant: string): Promise<void> {
  if (tenant === undefined || tenant === TENANT_DEFAULT) {
    logger.debug("default tenant, skip creation database");
    return;
  }

  if (await hasTenantDatabase(tenant)) {
    logger.info(
      "try to create database for tenant",
      tenant.green,
      "but its already existed"
    );
    return;
  }

  await runWithAdminConnection(async ds => {
    const databaseName = getTenantDatabaseName(tenant);
    logger.info("creating database", databaseName);
    // mysql 5.6 not support 'if not exists'
    await ds.query(`CREATE DATABASE ${databaseName}`);
    logger.info("database", databaseName, "created");
  });

}

/**
 * drop database for tenant
 * 
 * @param tenant 
 */
export async function dropDatabase(tenant: string) {
  logger.info("drop database for tenant", tenant);

  if (tenant === undefined) {
    throw new Error("tenant id must be provided for database drop");
  }

  await runWithAdminConnection(async ds => {
    const databaseName = getTenantDatabaseName(tenant);
    logger.info("drop database", databaseName);
    await ds.query(`DROP DATABASE ${databaseName}`);
    logger.info("drop database", databaseName, "successful");
  });
}



/**
 * sync tenant data model
 * 
 * @param tenant 
 * @param csn CSN, optional 
 * @returns 
 */
export async function syncTenant(tenant: string, csn?: CSN) {
  if (csn !== undefined) {
    await deploy(csn, tenant);
    return;
  }

  // if has tenant database & mtxs is enabled
  if (await hasTenantDatabase(tenant) && ("cds.xt.ModelProviderService" in cds.services)) {
    await deploy(await csn4(tenant), tenant);
    return;
  }

  await deploy(await _rawCSN(cds.db.model), tenant);
}

/**
 * deploy CSV (only) for tenant
 * 
 * @param tenant 
 */
export async function deployCSV(tenant?: string, csvList?: Array<string>) {
  // REVISIT: global model not suitable for extensibility
  await migrateData(getMySQLCredential(tenant), cds.model, csvList);
}

/**
 * deploy (migrate) database model into target tenant database
 * 
 * @param model plain CSN object
 * @param tenant tenant id
 * @returns 
 */
export async function deploy(model: CSN, tenant: string = TENANT_DEFAULT) {
  try {
    logger.info("migrating schema for tenant", tenant.green);

    if (tenant !== TENANT_DEFAULT) { await createDatabase(tenant); }
    const migrateOptions = await getDataSourceOption(tenant);
    const options = _options();
    if (
      options?.tenant?.deploy?.transparent === true &&
      getAdminTenantName() !== tenant // t0 need to use old way to deploy
    ) {
      logger.info("migrate with transparent approach");
      const migrations = migration_tool.parse(
        await fs.readFile(
          path.join(cds.root, "db/migrations.sql"),
          { encoding: "utf-8" }
        )
      );
      await migrate({ ...migrateOptions }, false, migrations);
    }
    else {
      const entities = csnToEntity(model);
      await migrate({ ...migrateOptions, entities });
    }

    if (options?.csv?.migrate !== false) {
      await deployCSV(tenant);
    }
    else {
      logger.info(
        "csv migration is disabled for tenant",
        tenant.green
      );
    }
    logger.info("migrate", "successful".green, "for tenant", tenant.green);
    return true;
  } catch (error) {
    logger.info("migrate", "failed".red, "for tenant", tenant.red, error);
    throw error;
  }
}

/**
 * get tenant 0 (admin tenant) name
 *
 * @returns
 */
export function getAdminTenantName() {
  return process.env.CDS_REQUIRES_MULTITENANCY_T0 ?? "t0";
}

/**
 * get (lower case) tables of target tenant
 * 
 * @param tenant 
 * @returns 
 */
export async function getTables(tenant?: string): Promise<Array<string>> {
  return runWithAdminConnection(async ds => {
    const tables = await ds.query(
      `SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE' AND TABLE_SCHEMA = ?`,
      [getTenantDatabaseName(tenant)],
    );
    return tables.map(({ TABLE_NAME }) => String(TABLE_NAME).toLowerCase());
  });
}

/**
 * get (lower case) columns of tables in target tenant
 * 
 * @param table table name
 * @param tenant tenant id
 * @returns 
 */
export async function getColumns(table: string, tenant?: string): Promise<Array<string>> {
  return runWithAdminConnection(async ds => {
    const records = await ds.query(
      `SELECT COLUMN_NAME FROM INFORMATION_SCHEMA.COLUMNS WHERE LOWER(TABLE_NAME) = ? AND TABLE_SCHEMA = ?`,
      [table.toLowerCase(), getTenantDatabaseName(tenant)],
    );
    return records.map(({ COLUMN_NAME }) => String(COLUMN_NAME).toLowerCase());
  });
}

/**
 * deploy admin tenant if required
 * 
 * @returns 
 */
export async function deployT0() {
  const t0 = getAdminTenantName();
  logger.info("deploy admin tenant", t0.green);
  const t0_csn = await cds.load(path.join(__dirname, "../mtxs/t0.cds"));
  logger.debug("t0 entities", Object.keys(t0_csn.definitions));
  await deploy(t0_csn, t0);
}

