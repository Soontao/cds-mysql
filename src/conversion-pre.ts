// pre conversion for INSERT/UPDATE (JSON -> DB)
import { DateTime } from "luxon";
import { MYSQL_DATE_TIME_FORMAT } from "./constants";
import { parseMysqlDate } from "./conversion";

/**
 * MySQL required wired formatted date time 'yyyy-MM-dd hh:mm:ss'
 * 
 * @param value 
 */
export function adaptToMySQLDateTime(value: string | Date) {
  if (value === null || value === undefined) {
    return null;
  }
  if (typeof value === "string") {

    // if is ISO datetime format
    if (value[10] === "T") {
      const datetime = DateTime.fromISO(value, { setZone: true });
      if (datetime.isValid) {
        return datetime.toUTC().toFormat(MYSQL_DATE_TIME_FORMAT);
      }
    }

    // if is valid mysql datetime value
    const datetime = parseMysqlDate(value);
    if (datetime.isValid) {
      return datetime.toFormat(MYSQL_DATE_TIME_FORMAT);
    }

  }

  // if is date object
  if (value instanceof Date) {
    return DateTime.fromJSDate(value).toFormat(MYSQL_DATE_TIME_FORMAT);
  }

  return null;

}

export const PRE_CONVERSION_MAP = new Map([
  ["cds.DateTime", adaptToMySQLDateTime],
  ["cds.Timestamp", adaptToMySQLDateTime]
]);