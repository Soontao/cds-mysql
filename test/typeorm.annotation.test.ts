import { cwdRequireCDS } from "cds-internal-tool";
import { csnToEntity } from "../src/typeorm";

describe("TypeORM Annotation Test Suite", () => {

  it("should support index for model", async () => {
    const cds = cwdRequireCDS();
    const csn = cds.parse.cdl(`
      @cds.typeorm.config : {indices : [{
        name    : 'NameIndex', // key name
        columns : ['Name'] // index fields
      }]}
      entity Demo {
        key ID: Integer;
            Name: String(255);
      }
    `);
    const entities =  csnToEntity(csn);
    expect(entities).toMatchSnapshot();
  });

});
