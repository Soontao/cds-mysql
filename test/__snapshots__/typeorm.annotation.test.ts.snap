// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeORM Annotation Test Suite should support index for model 1`] = `
[
  EntitySchema {
    "@instanceof": Symbol(EntitySchema),
    "options": {
      "columns": {
        "ID": {
          "name": "ID",
          "nullable": false,
          "primary": true,
          "type": "integer",
        },
        "Name": {
          "length": 255,
          "name": "Name",
          "nullable": true,
          "type": "nvarchar",
        },
      },
      "indices": [
        {
          "columns": [
            "Name",
          ],
          "name": "NameIndex",
        },
      ],
      "name": "Demo",
      "synchronize": true,
      "tableName": "Demo",
    },
  },
]
`;
